AI #1944: Refactor Attribute normative requirements to align with guidelines (#1981)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/null_handling.md

## Requirements

> Column conforming to NullHandling attribute MUST adhere to the following requirements:

Reviewer: Null Handling attribute is a dataset-level attribute.  Because of this, this needs to be rephrased.

```suggestion
FOCUS dataset columns MUST adhere to the following requirements:
```

Reviewer: Or:

```suggestion
Columns in a FOCUS dataset MUST adhere to the following requirements:
```

Author: Thanks for calling this out! Great point about NullHandling being a dataset-level attribute.

I'd like to suggest a slightly different approach - keep the requirements in the attribute AS-IS, but rephrase those dataset-level requirements.

---

### Dataset-level requirements

- **Dataset remains the subject**, but it’s clear that the **requirements apply to columns within the dataset**. This keeps the dataset-level context while making the target of the requirement explicit.

#### Current:
> * CostAndUsage MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.
> * CostAndUsage MUST conform to [NullHandling](#attributes.nullhandling) requirements.
> * CostAndUsage MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#attributes.datagenerator-calculatedsplitcostallocationhandling) requirements.
> * CostAndUsage MUST conform to [DatasetCompleteness](#attributes.datasetcompleteness) requirements.
> * CostAndUsage MUST conform to [DatasetConfiguration](#attributes.datasetconfiguration) requirements.

#### Option A:
> * CostAndUsage MUST conform to [ColumnHandling](#attributes.columnhandling) requirements for each column.
> * CostAndUsage MUST conform to [NullHandling](#attributes.nullhandling) requirements for each column.
> * CostAndUsage MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#attributes.datagenerator-calculatedsplitcostallocationhandling) requirements for each column.
> * CostAndUsage MUST conform to [DatasetCompleteness](#attributes.datasetcompleteness) requirements.
> * CostAndUsage MUST conform to [DatasetConfiguration](#attributes.datasetconfiguration) requirements.

#### Option B:
> * CostAndUsage MUST have its columns conform to [ColumnHandling](#attributes.columnhandling) requirements.
> * CostAndUsage MUST have its columns conform to [NullHandling](#attributes.nullhandling) requirements.
> * CostAndUsage MUST have its columns conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#attributes.datagenerator-calculatedsplitcostallocationhandling) requirements.
> * CostAndUsage MUST conform to [DatasetCompleteness](#attributes.datasetcompleteness) requirements.
> * CostAndUsage MUST conform to [DatasetConfiguration](#attributes.datasetconfiguration) requirements.

### Requirements listed in attributes

- At the same time, for requirements listed in the attribute, I’d prefer to:
  * Keep **Column** as the main subject, since each requirement is evaluated per column instance.
  * Use the **singular form** for subjects (as discussed in the RM meeting), because:
    - It clearly expresses that the requirement applies to an individual instance.
    - Singular naturally implies “for each instance.”
    - It aligns with common patterns and our RM discussion.

> Column conforming to NullHandling attribute MUST adhere to the following requirements:
>
> * FOCUS column MUST adhere to the following requirements:
>   * FOCUS column MUST use NULL for absent values when the FOCUS column is defined as nullable.
>   ...

---

@shawnalpay, what do you think?

Reviewer: @ijurica If you're saying the actor is a column, then the requirement should be written that way.

For example, in terms of English:

> CostAndUsage MUST conform to [ColumnHandling](#attributes.columnhandling) requirements for each column.

Is, in my opinion, MUCH clearer if written as:

> CostAndUsage columns MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.

It seems you're phrasing it this way and sidestepping the real actor for the sake of the Requirements Model implementation.  If so, then I disagree with that choice.

I would challenge you to write a guideline for this explaining why and how we would phrase it the way you have proposed.  I can assure you that I (and other FOCUS authors) will not get this right going forward because it does not read intuitively.

Reviewer 2: > CostAndUsage columns MUST conform to [ColumnHandling](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1981/files#attributes.columnhandling) requirements.

Is actually clearer from a modelling standpoint to be honest. Hiding "for each column." in the requirement means the entity class Is being hidden from the preamble. 

I vote we move forward with this option. 

if I was to say the "best way" from a modelling standpoint each column would have a link to ColumnHandling itself, but I do not like this for the specification. I think it's easier for us to have the classes up front in the preamble (\<dataset\> columns) and force the model reader to translate this to mean all columns get this requirement.

Author: Addressed in latest version.

Attaching snippets for reference:

> * InvoiceDetail MUST conform to [DatasetCompleteness](#attributes.datasetcompleteness) requirements.
> * InvoiceDetail MUST conform to [DatasetConfiguration](#attributes.datasetconfiguration) requirements.
> * InvoiceDetail MUST conform to [InvoiceHandling](#attributes.invoicehandling) requirements.
> * InvoiceDetail columns MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.
> * InvoiceDetail columns MUST conform to [NullHandling](#attributes.nullhandling) requirements.

> * CostAndUsage MUST conform to [DatasetCompleteness](#attributes.datasetcompleteness) requirements.
> * CostAndUsage MUST conform to [DatasetConfiguration](#attributes.datasetconfiguration) requirements.
> * CostAndUsage MUST conform to [DiscountHandling](#attributes.discounthandling) requirements.
> * CostAndUsage MUST conform to [InvoiceHandling](#attributes.invoicehandling) requirements.
> * CostAndUsage columns MUST conform to [ColumnHandling](#attributes.columnhandling) requirements.
> * CostAndUsage columns MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#attributes.datagenerator-calculatedsplitcostallocationhandling) requirements when the data generator supports data generator-calculated split cost allocation.
> * CostAndUsage columns MUST conform to [NullHandling](#attributes.nullhandling) requirements.

Resolving.

@shawnalpay, @mike-finopsorg, please feel free to reopen if I've missed something.

* FOCUS column MUST adhere to the following requirements:
  * FOCUS column defined as nullable MUST use NULL when no value is applicable.

>   * FOCUS column containing string values MUST NOT contain empty strings or placeholder strings (e.g., ""Not Applicable"") to indicate that no value is applicable.

Reviewer 3: Doubled double-quotes around "Not Applicable". The original spec used single double-quotes.

```suggestion
  * FOCUS column containing string values MUST NOT contain empty strings or placeholder strings (e.g., "Not Applicable") to indicate that no value is applicable.
```

  * FOCUS column containing numeric values MUST NOT contain placeholder numeric values (e.g., 0) to indicate that no value is applicable.
* Custom column MUST adhere to the following requirements:
  * Custom column defined as nullable SHOULD use NULL when no value is applicable.

>   * Custom column containing string values SHOULD NOT contain empty strings or placeholder strings (e.g., ""Not Applicable"") to indicate that no value is applicable.

Reviewer 3: Same doubled-quote issue on the Custom column line.

```suggestion
  * Custom column containing string values SHOULD NOT contain empty strings or placeholder strings (e.g., "Not Applicable") to indicate that no value is applicable.
```
